### PR TITLE
CRUD In registered customers

### PR DESCRIPTION
**Pull request by sdeoliveira-r**

### CODE REVIEW

File: atm-api/src/controllers/customers.ts [hunk 1]
```diff
@@ -65,3 +74,17 @@ export const updateCustomer = async (req: Request, res: Response) => {
...
+
+export const deleteCustomer = async (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+    const deletedCustomer = await deleteCustomerById(id);
+
+    return res.status(200).json({
+      message: 'Deleted!',
+      deletedCustomer,
+    });
```
Comment (davialcantaraa) [inline atm-api/src/controllers/customers.ts]: When deleting, no returned data is needed

File: atm-api/src/controllers/customers.ts [hunk 2]
```diff
@@ -35,10 +42,10 @@ export const deleteCustomer = async (req: Request, res: Response) => {
...
-    const { name } = req.body;
-    if (!name) {
+    const { name, email, phone } = req.body;
+    if (!name || !email || !phone) {
       return res.status(400).json({
-        message: 'Name is required!',
+        message: 'Name, email and phone is required!',
         name,
       });
     }
```
Comment (davialcantaraa) [inline atm-api/src/controllers/customers.ts]: A more useful validation needed

**Comment by sdeoliveira-r** [comment]
[[DEV] Registered Customers](https://trello.com/c/iDuFyauj/2-dev-registered-customers)

File: atm-api/src/controllers/customers.ts [hunk 3]
```diff
@@ -1,31 +1,38 @@
...
+export const getAllCustomers = async (req: Request, res: Response) => {
   try {
-    const { id } = req.params;
-    const deletedCustomer = await deleteCustomerById(id);
+    const customers = await getCustomers();
 
     return res.status(200).json({
-      message: 'Deleted!',
-      deletedCustomer,
+      message: 'All customers!',
```
Comment (davialcantaraa) [inline atm-api/src/controllers/customers.ts]: When returning data, no message needed

File: atm-api/src/controllers/customers.ts [hunk 4]
```diff
@@ -1,31 +1,38 @@
...
-      message: 'All customers!',
-      customers,
-    });
+    if (!customer) {
+      return res.status(404).json({
+        message: 'Customer not found!',
+      });
+    } else {
+      return res.status(200).json({
+        message: 'Customer found!',
```
Comment (davialcantaraa) [inline atm-api/src/controllers/customers.ts]: When returning data, no message needed

File: atm-api/src/routes/customers.ts [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
+//import { isAuthenticated } from '../middlewares/customers';
 
 const router = Router();
 
 router.get('/', getAllCustomers);
-router.delete('/delete/:id', isAuthenticated, deleteCustomer);
-router.patch('/:id', isAuthenticated, updateCustomer);
+router.get('/:email', getCustomer);
+router.patch('/:id', updateCustomer);
+router.delete('/:id', deleteCustomer);
```
Comment (davialcantaraa) [inline atm-api/src/routes/customers.ts]: PUT method preferred. Authentication middleware needed

File: atm-app/src/components/Customers/List.tsx [hunk 6]
```diff
@@ -9,43 +17,241 @@ interface Customer {
...
 const CustomersList: React.FC = () => {
+    const [selectedItem, setSelectedItem] = useState("list");
+    const [customer, setCustomer] = useState<any>(null);
     const [customers, setCustomers] = useState<Customer[]>([]);
+    const [selectedId, setSelectedId] = useState<string | null>(null);
+    const [selectedName, setSelectedName] = useState("");
+    const [selectedEmail, setSelectedEmail] = useState("");
+    const [selectedPhone, setSelectedPhone] = useState("");
+    const [showList, setShowList] = useState(false);
+    const [message, setMessage] = useState<string | null>(null);
```
Comment (davialcantaraa) [inline atm-app/src/components/Customers/List.tsx]: States with initialization values don't need to a type. 

https://www.typescriptlang.org/pt/docs/handbook/type-inference.html

File: atm-app/src/components/Customers/List.tsx [hunk 7]
```diff
@@ -9,43 +17,241 @@ interface Customer {
...
     const [error, setError] = useState<string | null>(null);
 
+    const showMessage = (msg: string) => {
+        setMessage(msg);
+        setTimeout(() => {
+            setMessage(null);
+        }, 3000);
+    };
+
     const handleCustomers = async () => {
```
Comment (davialcantaraa) [inline atm-app/src/components/Customers/List.tsx]: A more detailed function name is preferred. Exemple: `handleGetAllCustomers`, `handleGetCustomers`, etc

File: atm-app/src/components/Customers/List.tsx [hunk 8]
```diff
@@ -9,43 +17,241 @@ interface Customer {
...
+        if (selectedId !== null) {
+            const body = {
+                name: selectedName || "",
+                email: selectedEmail || "",
+                phone: selectedPhone || "",
+            };
+
+            if (!body.name || !body.email || !body.phone) {
+                showMessage("All fields are mandatory! Please try again.");
+            };
```
Comment (davialcantaraa) [inline atm-app/src/components/Customers/List.tsx]: A more useful validation needed

File: atm-app/src/components/Customers/List.tsx [hunk 9]
```diff
@@ -9,43 +17,241 @@ interface Customer {
...
+
             setError(null);
         } catch (error: any) {
             setError(error);
         }
     };
 
+    // reusable modal
+    const { isModalVisible, toggleModalVisibility } = useModal();
+    const modalContent: React.ReactNode = (
```
Comment (davialcantaraa) [inline atm-app/src/components/Customers/List.tsx]: This variable don't need a type. 

https://www.typescriptlang.org/pt/docs/handbook/type-inference.html

File: atm-app/src/services/customer/index.ts
Comment (davialcantaraa) [inline atm-app/src/services/customer/index.ts:1]: You can add a specific type for each function return.

Example:

```ts
// customers.d.ts

export type Customer {
    id: string,
    name: string,
    email: string
}
```

```ts
import { Customer } from "./customers.d.ts"
import { AxiosResponse } from "axios"

export const getCustomer = async (email: string): Promise<AxiosResponse<Customer>> => {
  try {
    const ENDPOINT = `customers/${email}`;
    const response = await axios.get(`${API_URL}${ENDPOINT}`);

    console.log(response.data);
    return response.data;
  } catch (error: any) {
    console.log(error);
    throw error.response.data;
  }
};
```

**Review by davialcantaraa — CHANGES_REQUESTED** [review]

**Review by sdeoliveira-r — COMMENTED** [review]
Viewed

**Review by sdeoliveira-r — COMMENTED** [review]
Fixes

**Review by sdeoliveira-r — COMMENTED** [review]
Viewed

**Pull request merged**